agent/consul/stream/event_publisher.go

```diff
@@ -116,34 +124,74 @@ func (e *EventPublisher) Run(ctx context.Context) {
 // publishEvent appends the events to any applicable topic buffers. It handles
 // any closeSubscriptionPayload events by closing associated subscriptions.
 func (e *EventPublisher) publishEvent(events []Event) {
-	eventsByTopic := make(map[Topic][]Event)
+	eventsByTopic := make(map[Topic]map[TopicKey][]Event)
 	for _, event := range events {
 		if unsubEvent, ok := event.Payload.(closeSubscriptionPayload); ok {
 			e.subscriptions.closeSubscriptionsForTokens(unsubEvent.tokensSecretIDs)
 			continue
 		}
 
-		eventsByTopic[event.Topic] = append(eventsByTopic[event.Topic], event)
+		byTopic, ok := eventsByTopic[event.Topic]
+		if !ok {
+			byTopic = make(map[TopicKey][]Event)
+			eventsByTopic[event.Topic] = byTopic
+		}
+
+		key := event.Payload.TopicKey()
+		byTopic[key] = append(byTopic[key], event)
 	}
 
 	e.lock.Lock()
 	defer e.lock.Unlock()
-	for topic, events := range eventsByTopic {
-		e.getTopicBuffer(topic).Append(events)
+	for topic, byKey := range eventsByTopic {
+		for key, events := range byKey {
+			// Note: bufferForPublishing returns nil if there are no subscribers for the
+			// given topic and key, in which case events will be dropped on the floor and
+			// future subscribers will catch up by consuming the snapshot.
+			if buf := e.bufferForPublishing(topic, key); buf != nil {
+				buf.Append(events)
+			}
+		}
```

Comment:
Note: IF it were possible to publish thousands of events in one call to Publish in the future, it might be worth avoiding the inner per-key loop here by checking there are even any subscribers at all for the topic first.

I think that would be a premature optimisation for now because:
 - We only ever publish a small number of events right now in practice.
 - Even with that check, in most clusters there is likely to be at least a few subscribers to every topic which means we have to loop by key anyway so we can't short circuit this inner loop.

tl;dr this is fine :)

Reply:
Good point! I think @dnephin's [idea](https://github.com/hashicorp/consul/pull/12080#discussion_r785216841) of including the `Topic` in `TopicKey` and making `topicBufs` a single-dimensional map means that we wouldn't need to think about this.